src/rgw/rgw_amqp_1.cc

```diff
@@ -0,0 +1,634 @@
+						conn->send(proton::message(message->message),message->cb);
+					}
+				}
+			}
+
+			// TODO
+			static void delete_message(const message_wrapper_t* message) {
+				delete message;
+			}
+
+		public:
+			Manager(size_t _max_connections,
+					size_t _max_inflight,
+					size_t _max_queue, 
+					CephContext* _cct) : 
+				max_connections(_max_connections),
+				max_inflight(_max_inflight),
+				max_queue(_max_queue),
+				connection_count(0),
+				stopped(false),
+				connections(_max_connections),
+				messages(max_queue),
+				queued(0),
+				dequeued(0),
+				cct(_cct),
+				// runner(&Manager::run, this),
+				// container_runner(&Manager::run_container, this) {
+				container_runner([&]() { container.run(); }) {
+					// The hashmap has "max connections" as the initial number of buckets, 
+					// and allows for 10 collisions per bucket before rehash.
+					// This is to prevent rehashing so that iterators are not invalidated 
+					// when a new connection is added.
+					connections.max_load_factor(10.0);
+					// give the runner thread a name for easier debugging
+					// this name should not be too long or the assertion fail
+					const auto rc = ceph_pthread_setname(container_runner.native_handle(),
+							"amqp1.0 manager");
+					ceph_assert(rc==0);
+				}
+
+			// non copyable
+			Manager(const Manager&) = delete;
+			const Manager& operator=(const Manager&) = delete;
+
+			// stop the main thread
+			void stop() {
+				stopped = true;
+			}
+
+			// disconnect from a broker
+			bool disconnect(connection_ptr_t& conn) {
+				if (!conn || stopped) {
+					return false;
+				}
+				conn->marked_for_deletion = true;
+				return true;
+			}
+
+			// TODO with params
+			connection_ptr_t connect(const std::string& url, bool use_ssl,
+					boost::optional<const std::string&> ca_location) {
+				if(stopped) {
+					ldout(cct, 1) << "AMQP_1 connect: manager is stopped" << dendl;
+					return nullptr;
+				}
+				// TODO: check url format parse
+				// TODO: max connections check
+				std::lock_guard<std::mutex> lock(connections_lock);
+
+				// TODO: find in the stale connectionlists
+
+				const auto conn = create_new_connection(container, url, cct, ca_location);
+
+				ceph_assert(conn);
+				++connection_count;
+				ldout(cct, 10) << "AMQP_1 connect: new connection is created. Total"
+					"connections: " << connection_count << dendl;
+				return connections.emplace(url, conn).first->second;
+			}
+
+			int publish(connection_ptr_t& conn, const std::string& topic, const
+					std::string& message) {
+				if(stopped) {
+					return RGW_AMQP_1_STATUS_MANAGER_STOPPED;
+				}
+				if(!conn || !conn->is_ok()) {
+					return RGW_AMQP_1_STATUS_CONNECTION_CLOSED;
+				}
+				// if(messages.push(new message_wrapper_t(conn, topic, message, nullptr)))
+				// 	{
+				// 		++queued;
+				// 		return RGW_AMQP_1_STATUS_OK;
+				// 	}
+				publish_internal(new message_wrapper_t(conn, topic, message, nullptr));
```

Comment:
the `publish()` function is invoked from another thread. it should not call `publish_internal()` that should just empty the message queue (and assumes it is invoked from the internal thread)